### DISCUSSION TITLE
Interactive view mode for Todo list

### DISCUSSION THREAD

**Discussion by tunbridgep** [gsantner/markor#1599] [Q&A] [answered]
Currently the workflow for dealing with Todo lists is very clunky.

First off, it starts in edit mode which shows us the syntax, which isn't that useful when we are actually in the middle of working.

Secondly, checking off an item involves either writing text or using the checkbox button at the bottom of the screen. View mode is largely useless as a result. But view mode looks a lot nicer, especially when using an md file, and looks a lot more like a traditional to-do list as it would in a dedicated Todo application

Instead of having to mess around with syntax, I feel like an "interactive view mode" would be a lot more useful for Todo lists.

When opening a Todo list, it would be in interactive view mode by default. This would give us a much nicer looking Todo list by default. Tapping on any item should toggle it's completed status on and off. Edit mode would still be needed for adding or editing entries.

This would be a Todo exclusive feature, as we don't want people accidentally updating documents elsewhere when in view mode since it's supposed to be read only.

The reason I'm asking for this is because tasks seem to be, for a lot of people, something they want to write upfront, then check off easily while in the middle of doing said tasks. Having to mess around in edit mode and constantly context switch as we complete tasks is time consuming, and just viewing it in edit mode all the time is really ugly and prone to accidental edits when we just want to check off tasks.

Vimwiki has a sort of similar feature where you can toggle checkboxes with ctrl-space regardless of whether or not you're in edit mode

**Comment by gsantner** [comment, marked as answer]
> When opening a Todo list, it would be in interactive view mode by default.

I won't make view mode editable. Period. If you try to reword it nicely, thats won't change a thing.

There are many todo.txt based apps out there, and all off them have some own custom GUI around it.  
Markor is different here, and it doesn't need to be just like the others. Thats what Markor makes different & unique and why people like it. If you like the other way, go for the other way.

As of view mode for todo.txt - yes it's not pretty. But at same time, in comparision to Markdown: There is zero interest by contributors to make it prettier. Markor is Open Source and everybody decides on their own if and how much time they give. For this one, nobody found time for it. People just always ask, rather working on it :smiley: . I also don't see much important either, the editor has proper highlight and works well. And - at the end of the day, it's still one line = one task, no things like videos, graphs, different headings. Thats where viewing stuff differently gets interesting, and what users do with Markdown and Zim. But not with todo.txt, the format is what it is - simple and lightweight.

[Issue # 859 (todo.txt: view mode - Fancy grouping, sort, search etc) at gsantner/markor](https://github.com/gsantner/markor/issues/859 )

**Comment by harshad1** [comment]
Fwiw, if you touch the title, you will get some options for more advanced filtering. You can, for example, search by project and then multi-select etc.

**Reply by xkcd386at** [reply]
yes but even the advanced filtering shows up inside a search box, and the only *change* allowed there is to mark the task complete

if you need any other type of change on a filtered view, you long click on the search result, and you're back in the main editor -- where the item you want is surrounded by unrelated items that just happen to be physically (in the text file on disk) near the one you want, and the *other* related items are probably elsewhere

it all comes down to one basic fact: a todo file is *fundamentally* different from a markdown file in terms of how "connected" the lines are to adjacent lines.  In a markdown file, line 10 is *related* to lines 9 and 11 in some manner, and you can't simply move them around.  In a todo file, line 10 may be connected to line 23 and 35, not to lines 9 and 11

since todo is fundamentally different from markdown, the whole concept of "view mode versus edit mode" breaks down, because for a todo file, it is reasonable to see a filtered set (which may consist of lines, 10, 23, and 35 in the physical file) *without* the distraction of everything else, and want to do *everything* to that set that you would be able to do to the full file

gsantner says all other todo apps have some custom GUI, and this is the reason -- they don't treat a todo file like a markdown file

I'm a huge fan of markor (markor <-> syncthing <-> vim has been my mantra response on many reddit threads for a couple of years now).  But only for the markdown part.  For the todo part, I have *reluctantly* moved to SimpleTask.  **Before I did that, I went through all the discussion items on todo**, and reluctantly came to the conclusion that this is not a feature issue that can be solved with code; it's a philosophical difference of opinion

thanks for listening

**Reply by gsantner** [reply]
> they don't treat a todo file like a markdown file

todo.txt files are **not** treated like markdown files at Markor. they are treated as todo.txt plaintext files with todo.txt specific implementation. again, Markor is focused on text editing, similary to vim in that regard. And if you want something different with a GUI like SimpleNote and many others have then use it, thats it.

**Reply by harshad1** [reply]
> yes but even the advanced filtering shows up inside a search box, and the only change allowed there is to mark the task complete

You can select one or more tasks in the advanced filtering view and on hitting ok, _all selected tasks_ will be highlighted in the regular edit view. You can add a context, change priority or due date for all of the selected tasks.

For example, at the start of my work day, I sort by priority, filter by +work and select all the tasks I want to get done. Back in the text editor I can then apply a due date of today to all these tasks.

Necessarily, selecting multiple disconnected tasks will reorder the file in order for them all to be selected at the same time (all selected tasks will be moved to the top). (The file is not changed if you select one or more contiguous taks)

I'll also point out that one of the great things about todo.txt in a plain file is that one can use multiple apps seamlessly. I will jump between markor and simpletask. I am (slowly) working on bring more and more functionality into markor's todo implementation while keeping it text file centric

**Reply by xkcd386at** [reply]
> You can select one or more tasks in the advanced filtering view and on hitting ok, _all selected tasks_ will be highlighted in the regular edit view. You can add a context, change priority or due date for all of the selected tasks.

oh that is nice; thank you.  I thought those check boxes were to mark the task complete (remember I said "and the only change allowed there is to mark the task complete"?)

not to hijack the thread but this brings me to another nit: the features are not discoverable.  Long click the *keyboard* icon to do something?  Long click the title to jump top/bottom?  I could only find them by laboriously going through the discussion responses (mostly yours, tbh!).  If they are documented elsewhere, I did not find them

but I'm happy to hear what you're doing.  Once "rec:" is implemented, I'll come back to using the todo part also

**Comment by harshad1** [comment]
rec, threshold, and hidden are probably things I'll implement at some stage.

Regarding discoverability, you're right. We need to document these a lot more and I haven't found the time / inclination to do it. If you're able to add to the readme / docs, that would be great imo.

Jul. 25, 2022 04:42:37 xkcd386at ***@***.***>:

**Reply by xkcd386at** [reply]
the readme does not have a section for documentation, and there is no separate doc directory or similar -- in fact the main project repository has no documentation at all, so I have no idea where to add this.

however, if you know where to add it, a good starting point would be:

----8<----

### navigating the document quickly

edit mode:

-   short press the title to get a drop-down table of contents, pick from there to go
-   long press the "keyboard" icon in the bottom toolbar to go to the top
    -   (if you're already at the top it goes to the bottom)

view mode:

-   long press the title to jump to the top of the document
    -   (but if you are already at the top it goes to about the middle of the document, NOT the bottom.  Bug??)

----8<----

there was one more piece where long press was discussed, about tables, so someone familiar with that feature could add that

**Reply by gsantner** [reply]
the github discsussion, the FAQ and the website are there for information. having a fourth point of it means another point to maintain, keep updated, and add things to. Recently it has shown to me that github discussions work pretty well though for many things. And at the end there are things that stand in every blog post, in the FAQ, in the project page, here on github discussions, and still I get the same question per private mail. So what I'm trying is to build up public knowledge here and especially to those most interesting topics users are interested in ... and thats what they can do here, ask themself.

I'm already spending most of time I can give (from my free time) on maintenance in this project. Which means less time for improving the code and the project itself. If you are interested in better documentation, start making one/update it, and keep it maintained, continuously. The project needs people.

**Reply by xkcd386at** [reply]
@gsantner:

I was responding to @harshad1, who suggested I should document some of the things I only discovered going through pretty much ALL the discussion items

as for your response, take a look at the three short items in my list and tell me where you expect your users *ideally* should find them.  Do you expect every single user to crawl all the discussions looking for "how to navigate to top of document" or would you rather they see it in some obvious place?

If you add a "doc" section to the project, it can only help you.  By keeping the documentation in your personal site, you're making it more difficult for yourself.  And then you wonder why they keep asking you in personal email.

Anyway, your project, your call; I don't have anything else to add.

**Reply by gsantner** [reply]
> If you add a "doc" section to the project, it can only help you. By keeping the documentation in your personal site, you're making it more difficult for yourself. And then you wonder why they keep asking you in personal email.

I didn't disagree with that idea. It's just, it needs **somebody** to write there and keep it updated. Which should ideally be somebody else than harshad and me. And again, even if it's there, it doesn't mean it extinguishes questions. People still will ask and want explanation (shorter , longer, more precise, less precise... depending on needs), even if it's written somewhere already. You can wonderful describe every aspect of a thing in 10 lines of text. But people often just want it short in a few words. 

> By keeping the documentation in your personal site, you're making it more difficult for yourself.

I tried with public writing posts/faq/description and asked people to join writing. Nobody wrote/added a word to that. So whatever you try to say here, my "personal site" also IS the project official web site, whether people want to help writing there or not. And so far people rather not want to help here. BUT they rather write here on GitHub discussions. And thats also what other people can find.

So long story short, if you want to improve something, please help with writing. If you just want to ask for a change, but don't help..thanks, but no thanks. People want a lot of things, but lifetime and time you can spend after work and in weekends is limited.
